File: Java/JsonlTraceSink.java
package mu.cse.rag.clean;

import java.io.FileWriter;
import java.io.IOException;
import java.time.LocalDateTime;

public class JsonlTraceSink implements TraceListener {

    private final String filePath; //JSONL file path

    public JsonlTraceSink(String filePath) {
        this.filePath = filePath;
    }

    @Override
    @SuppressWarnings("CallToPrintStackTrace")
    public void onEvent(TraceEvent event) { //Overrides the method
        try (FileWriter filewriter = new FileWriter(filePath, true)) {
            String json = String.format(
                    "{\"stage\":\"%s\",\"input\":\"%s\",\"output\":\"%s\",\"durationMs\":%d,\"timestamp\":\"%s\"}",
                    escape(event.stage), //Replaces double quotes with \"
                    escape(event.inputSummary), //Replaces double quotes with \"
                    escape(event.outputSummary), //Replaces double quotes with \"
                    event.durationMs, //Time taken
                    LocalDateTime.now().toString()
            );
            filewriter.write(json + "\n");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private String escape(String s) {
        return s.replace("\"", "\\\"");
    }
}

File: Python/JsonlTraceSink.py
import json
from datetime import datetime
from typing import Any

from .TraceListener import TraceListener
from .TraceEvent import TraceEvent


class JsonlTraceSink(TraceListener):
    def __init__(self, file_path: str):
        self.file_path = file_path

    def on_event(self, event: TraceEvent) -> None:
        record: dict[str, Any] = {
            "stage": event.stage,
            "input": event.inputSummary,
            "output": event.outputSummary,
            "durationMs": event.durationMs,
            "timestamp": datetime.now().isoformat(),
        }
        try:
            with open(self.file_path, 'a', encoding='utf-8') as f:
                f.write(json.dumps(record, ensure_ascii=False) + "\n")
        except Exception:
            pass

